SnapshotResiliencyTests testSnapshotWithNodeDisconnects and testSnapshotPrimaryRelocations failures

**Build scan**:
https://gradle-enterprise.elastic.co/s/puwxv532d7p5u

**Repro line**:
```
 ./gradlew ':server:test' --tests "org.elasticsearch.snapshots.SnapshotResiliencyTests.testSnapshotWithNodeDisconnects" \
  -Dtests.seed=AB11F5CEE90655C4 \
  -Dtests.security.manager=true \
  -Dtests.locale=ar-SY \
  -Dtests.timezone=Pacific/Pitcairn \
  -Druntime.java=11
```

**Reproduces locally?**:
Yes 😲 

**Applicable branches**:
Master

**Failure history**:
Once or twice a day since 08/07/2020

https://build-stats.elastic.co/app/kibana#/discover?_g=(refreshInterval:(pause:!t,value:0),time:(from:now-7d,mode:quick,to:now))&_a=(columns:!(branch,class,stacktrace),index:e58bf320-7efd-11e8-bf69-63c8ef516157,interval:auto,query:(language:lucene,query:'SnapshotResiliencyTests%20testSnapshotWithNodeDisconnect'),sort:!(time,desc))

**Failure excerpt**:
```
java.lang.AssertionError: Condition wasn't fulfilled.
	at __randomizedtesting.SeedInfo.seed([AB11F5CEE90655C4:A5DE5A06F3C06E04]:0)
	at org.junit.Assert.fail(Assert.java:88)
	at org.elasticsearch.snapshots.SnapshotResiliencyTests.runUntil(SnapshotResiliencyTests.java:1163)
	at org.elasticsearch.snapshots.SnapshotResiliencyTests.verifyReposThenStopServices(SnapshotResiliencyTests.java:280)
```

'Condition wasn't fulfilled' <- you are such a disappointment

--- a/server/src/internalClusterTest/java/org/elasticsearch/snapshots/DedicatedClusterSnapshotRestoreIT.java
+++ b/server/src/internalClusterTest/java/org/elasticsearch/snapshots/DedicatedClusterSnapshotRestoreIT.java
@@ -1250,6 +1250,19 @@ public class DedicatedClusterSnapshotRestoreIT extends AbstractSnapshotIntegTest
         assertThat(snapshotResponse.get().getSnapshotInfo().state(), is(SnapshotState.FAILED));
     }
 
+    public void testPartialSnapshotAllShardsMissing() throws Exception {
+        internalCluster().startMasterOnlyNode();
+        final String dataNode = internalCluster().startDataOnlyNode();
+        final String repoName = "test-repo";
+        createRepository(repoName, "fs");
+        createIndex("some-index");
+        stopNode(dataNode);
+        ensureStableCluster(1);
+        final CreateSnapshotResponse createSnapshotResponse = client().admin().cluster().prepareCreateSnapshot(repoName, "test-snap")
+                .setPartial(true).setWaitForCompletion(true).get();
+        assertThat(createSnapshotResponse.getSnapshotInfo().state(), is(SnapshotState.PARTIAL));
+    }
+
     private long calculateTotalFilesSize(List<Path> files) {
         return files.stream().mapToLong(f -> {
             try {

--- a/server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java
+++ b/server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java
@@ -314,7 +314,7 @@ public class SnapshotsService extends AbstractLifecycleComponent implements Clus
                     logger.info("snapshot [{}] started", snapshot);
                     listener.onResponse(snapshot);
                 } finally {
-                    if (newEntry.state().completed() || newEntry.shards().isEmpty()) {
+                    if (newEntry.state().completed() || completed(newEntry.shards().values())) {
                         endSnapshot(newEntry, newState.metadata(), repositoryData);
                     }
                 }

--- a/server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java
+++ b/server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java
@@ -294,8 +294,8 @@ public class SnapshotsService extends AbstractLifecycleComponent implements Clus
                 }
                 newEntry = new SnapshotsInProgress.Entry(
                         new Snapshot(repositoryName, snapshotId), request.includeGlobalState(), request.partial(),
-                        State.STARTED, indexIds, dataStreams, threadPool.absoluteTimeInMillis(), repositoryData.getGenId(), shards,
-                        null, userMeta, version);
+                        completed(shards.values()) ? State.SUCCESS : State.STARTED, indexIds, dataStreams,
+                        threadPool.absoluteTimeInMillis(), repositoryData.getGenId(), shards, null, userMeta, version);
                 final List<SnapshotsInProgress.Entry> newEntries = new ArrayList<>(runningSnapshots);
                 newEntries.add(newEntry);
                 return ClusterState.builder(currentState).putCustom(SnapshotsInProgress.TYPE,
@@ -314,7 +314,7 @@ public class SnapshotsService extends AbstractLifecycleComponent implements Clus
                     logger.info("snapshot [{}] started", snapshot);
                     listener.onResponse(snapshot);
                 } finally {
-                    if (newEntry.state().completed() || completed(newEntry.shards().values())) {
+                    if (newEntry.state().completed()) {
                         endSnapshot(newEntry, newState.metadata(), repositoryData);
                     }
                 }

--- a/server/src/test/java/org/elasticsearch/snapshots/SnapshotResiliencyTests.java
+++ b/server/src/test/java/org/elasticsearch/snapshots/SnapshotResiliencyTests.java
@@ -363,7 +363,6 @@ public class SnapshotResiliencyTests extends ESTestCase {
         assertEquals(0, snapshotInfo.failedShards());
     }
 
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/59384")
     public void testSnapshotWithNodeDisconnects() {
         final int dataNodes = randomIntBetween(2, 10);
         final int masterNodes = randomFrom(1, 3, 5);

